packages/store-sync/src/dozer/getSnapshot.ts

```diff
@@ -0,0 +1,88 @@
+import { LogFilter, SyncFilter, TableQuery } from "./common";
+import { Hex } from "viem";
+import { StorageAdapterBlock, SyncFilter as LegacyLogFilter } from "../common";
+import { fetchRecordsSql } from "./fetchRecordsSql";
+import { recordToLog } from "../recordToLog";
+import { getSnapshot as getSnapshotLogs } from "../getSnapshot";
+import { bigIntMin, isDefined } from "@latticexyz/common/utils";
+
+export type GetSnapshotArgs = {
+  dozerUrl: string;
+  storeAddress: Hex;
+  filters?: SyncFilter[];
+  startBlock?: bigint;
+  chainId: number;
+};
+
+export type GetSnapshotResult = {
+  initialBlockLogs: StorageAdapterBlock;
+};
+
+export async function getSnapshot({
+  dozerUrl,
+  storeAddress,
+  filters,
+  startBlock = 0n,
+  chainId,
+}: GetSnapshotArgs): Promise<GetSnapshotResult> {
+  // We execute the list of provided SQL queries for hydration. For performance
+  // reasons the queries are not executed against a fixed block height, but against
+  // the latest state. We therefore pass the min block number of all query results
+  // as overall block number. This means some logs will be re-fetched again during
+  // the hydration process, but after the hydration is complete, the state will be
+  // correct. Intermediate state updates during hydration might be incorrect (for
+  // partial updates), so we only notify consumers of state updates after the
+  // initial hydration is complete.
+
+  const sqlFilters = filters ? (filters.filter((filter) => "sql" in filter) as TableQuery[]) : [];
+
+  const fetchLogs = async (): Promise<StorageAdapterBlock | undefined> => {
+    // Fetch the tables without SQL filter from the snapshot logs API for better performance.
+    const logsFilters =
+      filters &&
+      filters
+        .filter((filter) => !("sql" in filter))
+        .map((filter) => {
+          const { table, key0, key1 } = filter as LogFilter;
+          return { tableId: table.tableId, key0, key1 } as LegacyLogFilter;
+        });
+
+    if (logsFilters && logsFilters.length === 0) {
+      return undefined;
+    }
+
+    return getSnapshotLogs({
+      chainId,
+      address: storeAddress,
+      filters: logsFilters,
+      indexerUrl: dozerUrl,
+    });
+  };
+
+  const fetchSql = (): Promise<StorageAdapterBlock | undefined>[] => {
+    return sqlFilters.map(async (filter) => {
+      const result = await fetchRecordsSql({ dozerUrl, storeAddress, queries: [filter] });
```

Comment:
more of a stylistic thing but I wonder if this would be more readable

```ts
  const fetchRecords = async (query: TableQuery): Promise<StorageAdapterBlock | undefined> => {
    const result = await fetchRecordsSql({ dozerUrl, storeAddress, queries: [filter] });
```

```ts
const results = (await Promise.all([fetchLogs(), ...sqlFilters.map(fetchRecords)])).filter(isDefined);
```